quinn-proto/src/tests/mod.rs

```diff
@@ -1177,6 +1177,39 @@ fn keep_alive() {
     }
 }
 
+#[test]
+fn cid_retirement() {
+    let _guard = subscribe();
+    let mut pair = Pair::default();
+    let (client_ch, server_ch) = pair.connect();
+
+    // Server retires current active remote CIDs
+    pair.server_conn_mut(server_ch).mock_retire_cid(1, 1);
+    pair.drive();
+    // Any unexpected behavior may trigger TransportError::CONNECTION_ID_LIMIT_ERROR
+    assert!(!pair.client_conn_mut(client_ch).is_closed());
+    assert!(!pair.server_conn_mut(server_ch).is_closed());
+    assert_matches!(pair.client_conn_mut(client_ch).active_rem_cid_seq(), 1);
```

Comment:
In this test case, 
1. Client initially holds active remote cids 0,1,2,3,4.
2. Server actively retire cid 0 and send a NEW_CONNECTION_ID frame to client.
3. When client processes this frame [here](https://github.com/quinn-rs/quinn/pull/871/files#diff-423ed266e9aaf5fcae0fa8b85ca382f495528c18cd7c890cef1cf438221f2ef6R2398), the offset is not changed (remains to be 1). This is current since we store 1,2,3,4 in CidQueue while we actually need to retire 0. However, later on when the new cid is inserted [here](https://github.com/quinn-rs/quinn/pull/871/files#diff-423ed266e9aaf5fcae0fa8b85ca382f495528c18cd7c890cef1cf438221f2ef6R2431), since offset is still 1, the index is calculated as 4 [here](https://github.com/quinn-rs/quinn/pull/871/files#diff-e07d31049906f88a397dc54fc600e122da5f719a4a91599e5dd513d321d2077fR41) which triggers ExceedsLimit error. In fact, client simply wants to change cid from 0,1,2,3,4 to 1,2,3,4,5 and there is no violation

Reply:
If i checkout commit `55168f3` and run this unit test. Here is the output
```
Oct 16 10:39:06.204 TRACE client:send{space=Data pn=2}: quinn_proto::connection: NEW_CONNECTION_ID sequence=5 id=a97e4574bfe160fd
Oct 16 10:39:06.204 TRACE client: quinn_proto::connection: sending 62 byte datagram
Oct 16 10:39:06.204 TRACE server: quinn_proto::connection: got Data packet (62 bytes) from [::1]:44433 using id 4ba0ad1102d392b7
Oct 16 10:39:06.204 TRACE server:recv{space=Data pn=2}: quinn_proto::connection: authenticated
Oct 16 10:39:06.204 TRACE server:recv{space=Data pn=2}:frame{ty=NEW_CONNECTION_ID}: quinn_proto::connection: sequence=5 id=a97e4574bfe160fd
Oct 16 10:39:06.204 TRACE server:send{space=Data pn=2}: quinn_proto::connection: ACK
Oct 16 10:39:06.204 TRACE server: quinn_proto::connection: sending 34 byte datagram
Oct 16 10:39:06.205 TRACE client: quinn_proto::connection: got Data packet (34 bytes) from [::1]:4433 using id a3710748a204985b
Oct 16 10:39:06.205 TRACE client:recv{space=Data pn=2}: quinn_proto::connection: authenticated
Oct 16 10:39:06.205 TRACE server:send{space=Data pn=3}: quinn_proto::connection: ACK
Oct 16 10:39:06.205 TRACE server:send{space=Data pn=3}: quinn_proto::connection: NEW_CONNECTION_ID sequence=5 id=c17e354033199672
Oct 16 10:39:06.205 TRACE server: quinn_proto::connection: sending 62 byte datagram
Oct 16 10:39:06.206 TRACE client: quinn_proto::connection: got Data packet (62 bytes) from [::1]:4433 using id a3710748a204985b
Oct 16 10:39:06.206 TRACE client:recv{space=Data pn=3}: quinn_proto::connection: authenticated
Oct 16 10:39:06.206 TRACE client:recv{space=Data pn=3}:frame{ty=NEW_CONNECTION_ID}: quinn_proto::connection: sequence=5 id=c17e354033199672
Oct 16 10:39:06.206 DEBUG client: quinn_proto::connection: closing connection due to transport error: the number of connection IDs provided by the peer exceeds the advertised active_connection_id_limit
Oct 16 10:39:06.206 TRACE client: quinn_proto::connection: connection closed
Oct 16 10:39:06.206 TRACE client:send{space=Data pn=3}: quinn_proto::connection: sending CONNECTION_CLOSE
Oct 16 10:39:06.206 TRACE client: quinn_proto::connection: sending 30 byte datagram
Oct 16 10:39:06.206 TRACE server: quinn_proto::connection: got Data packet (30 bytes) from [::1]:44433 using id 4ba0ad1102d392b7
Oct 16 10:39:06.207 TRACE server:recv{space=Data pn=3}: quinn_proto::connection: authenticated
Oct 16 10:39:06.208 TRACE server: quinn_proto::connection: connection closed
Oct 16 10:39:06.208 TRACE server:send{space=Data pn=4}: quinn_proto::connection: sending CONNECTION_CLOSE
Oct 16 10:39:06.208 TRACE server: quinn_proto::connection: sending 30 byte datagram
Oct 16 10:39:06.208 TRACE client: quinn_proto::connection: got Data packet (30 bytes) from [::1]:4433 using id a3710748a204985b
Oct 16 10:39:06.209 TRACE client:recv{space=Data pn=4}: quinn_proto::connection: draining
Oct 16 10:39:06.209 TRACE quinn_proto::tests::util: advancing to Instant { t: 305032168058444 } for client
Oct 16 10:39:06.209 TRACE client: quinn_proto::connection: timeout timer=Close
Oct 16 10:39:06.209 TRACE server: quinn_proto::connection: timeout timer=Close
```

Reply:
To fix this issue, [this code](https://github.com/quinn-rs/quinn/pull/871/files#diff-423ed266e9aaf5fcae0fa8b85ca382f495528c18cd7c890cef1cf438221f2ef6R2404-R2412) has to move before CidQueue insert to allow we get any replacement of cid from CidQueue and reset the offset value.

if retire_prior_to field invalidates everything (including both rem_cid and rem_cids), the execution flow will just to [this](https://github.com/quinn-rs/quinn/pull/871/files#diff-423ed266e9aaf5fcae0fa8b85ca382f495528c18cd7c890cef1cf438221f2ef6R2446-R2455) where we assign cid in current frame to `rem_cid`

[This](https://github.com/quinn-rs/quinn/pull/871/files#diff-423ed266e9aaf5fcae0fa8b85ca382f495528c18cd7c890cef1cf438221f2ef6R2446-R2455) handles the corner case of scenario where retire_prior_to invalidates everything. 
Consider the following scenario: 
1. Peer A has 0,1,2,3,4 as valid cids. 
2. Peer B wants to provide 2 new cids: cid 5 and 6 and invalidates everything it previously issued. To do that, peer B set retire_prior_to to 5 and send these two NEW_CONNECTION_ID frame.
3. As we cannot control on the other end which frame is processed first, it is possible peer A first deals with new cid 6 and then 5. As a result, [This](https://github.com/quinn-rs/quinn/pull/871/files#diff-423ed266e9aaf5fcae0fa8b85ca382f495528c18cd7c890cef1cf438221f2ef6R2446-R2455) is needed to swap cid and make sure `rem_cid` always has the smallest sequence number.